neuralpredictors/training/tracking.py

```diff
@@ -146,3 +153,360 @@ def asdict(self, time_key="time", make_copy=True):
+        The entry reflects the default value saved in `objectives`.
+        Args:
+            log: sub dictionary to add the objectives to
+            objectives: dictionary of default values
+        """
+        for key, objective in objectives.items():
+            if isinstance(objective, dict):
+                self._initialize_epoch(log[key], objective)
+            elif not callable(objective):
+                while len(log[key]) <= self.epoch:
+                    log[key].append(objective)
+
+    def _log_objective_value(self, value: float, log: Dict, key: Tuple[str, ...] = ()):
+        """
+        Recursively walk through the log dictionary to get to follow `key`.
+        When on lowest level: add `value` to entry at current epoch.
+        Args:
+            value: objective value to log
+            log: log subdict to add value to
+            key: key for where `value` is saved
+        """
+        if len(key) > 1:
+            self._log_objective_value(value, log[key[0]], key[1:])
+        else:
+            log[key[0]][self.epoch] += value
+
+    def _log_objective_callables(self, log: Dict, objectives: Dict):
+        """
+        Log objectives that are specified as callables
+        Disclaimer: this is not very well tested!
+        Args:
+            log:
+            objectives:
+        """
+        for key, objective in objectives.items():
+            if isinstance(objective, dict):
+                self._log_objective_callables(log[key], objective)
+            elif callable(objective):
+                log[key].append(objective())
+
+    def _normalize_log(self, log: [Dict, List]) -> [Dict, np.array]:
+        """
+        Recursively go through the log and normalize the entries that
+        have `"normalization"` information on the same level.
+        Args:
+            log: subdict on which to apply normalization or list (if on lowest level)
+
+        Returns:
+            Normalized log dictionary that has numpy arrays on the lowest level
+
+        """
+        if isinstance(log, dict):
+            n_log = {}
+            norm = None
+            for key, l in log.items():
+                res = self._normalize_log(l)  # to turn into arrays
+                if key == "normalization":
+                    assert isinstance(res, np.ndarray)
+                    norm = res
+                else:
+                    n_log[key] = res
+            if norm is not None:
+                nonzero_start = (norm != 0).argmax(axis=0)
+                norm = norm[nonzero_start:]
+                for key, l in n_log.items():
+                    l = l[nonzero_start:]
+                    if isinstance(l, np.ndarray):
+                        n_log[key] = l / np.where(norm > 0, norm, np.ones_like(norm))
+            return n_log
+        else:
+            return np.array(log)
+
+    def _gather_log(self, log: Dict, key: Tuple[str, ...], index: int = -1) -> Dict:
```

Comment:
Judging from the code the type of `log` can be `List` (or more generally `Sequence`) as well as `Dict` (or `Mapping`) but the signature only specifies `Dict`.